segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/chunklayer/ChunkStorageManager.java

```diff
@@ -0,0 +1,1375 @@
+                ChunkMetadata sourceFirstChunk = (ChunkMetadata) txn.get(sourceSegmentMetadata.getFirstChunk());
+
+                if (targetLastChunk != null) {
+                    targetLastChunk.setNextChunk(sourceFirstChunk.getName());
+                    txn.update(targetLastChunk);
+                } else {
+                    if (sourceFirstChunk != null) {
+                        targetSegmentMetadata.setFirstChunk(sourceFirstChunk.getName());
+                        txn.update(sourceFirstChunk);
+                    }
+                }
+
+                // Update segments's last chunk to point to the sources last segment.
+                targetSegmentMetadata.setLastChunk(sourceSegmentMetadata.getLastChunk());
+
+                // Update the length of segment.
+                targetSegmentMetadata.setLastChunkStartOffset(targetSegmentMetadata.getLength() + sourceSegmentMetadata.getLastChunkStartOffset());
+                targetSegmentMetadata.setLength(targetSegmentMetadata.getLength() + sourceSegmentMetadata.getLength() - sourceSegmentMetadata.getStartOffset());
+
+                txn.update(targetSegmentMetadata);
+                txn.delete(sourceSegment);
+
+                // Finally defrag immediately.
+                if (chunkStorage.supportsConcat() && null != targetLastChunk ) {
+                    defrag(txn, targetSegmentMetadata, targetLastChunk.getName(), null);
+                }
+
+                targetSegmentMetadata.checkInvariants();
+
+                // Finally commit transaction.
+                txn.commit();
+
+                Duration elapsed = timer.getElapsed();
+                log.debug("concat - target={} source={} offset={} latency={}.", targetHandle.getSegmentName(), sourceSegment, offset, elapsed.toMillis());
+                LoggerHelpers.traceLeave(log, "concat", traceId, targetHandle, offset, sourceSegment);
+
+                // Update the read index.
+                cachedReadIndex.remove(sourceSegment);
+
+            } catch (StorageMetadataVersionMismatchException ex) {
+                throw ex;
+            } catch (StorageMetadataWritesFencedOutException ex) {
+                throw new StorageNotPrimaryException(targetSegmentName, ex);
+            }
+
+            return null;
+        });
+    }
+
+    /**
+     * Defragments the combined chunks after concatenation.
+     * @param txn Active {@link MetadataTransaction}.
+     * @param segmentMetadata {@link SegmentMetadata} for the segment to defrag.
+     * @param startChunkName Name of the first chunk to start defragmentation.
+     * @param lastChunkName Name of the last chunk before which to stop defragmentation. (last chunk is not concatenated).
+     * @throws Exception In case of any errors.
+     */
+    @VisibleForTesting
+    private void defrag(MetadataTransaction txn, SegmentMetadata segmentMetadata, String startChunkName, String lastChunkName) throws Exception {
```

Comment:
You need to explain in great detail what this does. Please put it in the Javadoc (instead of a comment inside the method).

Reply:
sure.